### PR TITLE
Tentative fix for the Pyrmont Eth1 voting issue

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

File: beacon_chain/rpc/debug_api.nim [hunk 1]
```diff
@@ -20,3 +22,15 @@ proc installDebugApiHandlers*(rpcServer: RpcServer, node: BeaconNode) =
   rpcServer.rpc("get_v1_debug_beacon_heads") do (
       stateId: string) -> seq[tuple[root: Eth2Digest, slot: Slot]]:
     return node.chainDag.heads.mapIt((it.root, it.slot))
+
+  rpcServer.rpc("get_v1_debug_eth1_chain") do () -> seq[Eth1Block]:
```
Comment (arnetheduck) [inline beacon_chain/rpc/debug_api.nim:26]: I assume these are not part of the standard upstream api so they should go into `nimbus_api`

Comment (arnetheduck) [reply beacon_chain/rpc/debug_api.nim:26]: also, it's important to add a manual entry with an example invocation for every api

**Comment by arnetheduck** [comment]
lots of lengthy lockups when processing the eth1 blocks - also, while it's processing pre-genesis blocks, it returns a pre-genesis eth1 vote

File: beacon_chain/rpc/debug_api.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/rpc/debug_api.nim:26]: crashes when no web3 url is specified

File: beacon_chain/rpc/debug_api.nim [hunk 2]
```diff
@@ -20,3 +22,15 @@ proc installDebugApiHandlers*(rpcServer: RpcServer, node: BeaconNode) =
...
+    return mapIt(node.eth1Monitor.blocks, it)
+
+  rpcServer.rpc("get_v1_debug_eth1_proposal_data") do () -> BlockProposalEth1Data:
+    let
+      wallSlot = node.beaconClock.now.slotOrZero
+      head = node.doChecksAndGetCurrentHead(wallSlot)
+
+    node.chainDag.withState(node.chainDag.tmpState, head.atSlot(wallSlot)):
```
Comment (arnetheduck) [inline beacon_chain/rpc/debug_api.nim:34]: add eth1data to epochref and avoid the state fetch?

File: beacon_chain/eth1_monitor.nim [hunk 3]
```diff
@@ -180,8 +200,11 @@ func addBlock(eth1Chain: var Eth1Chain, newBlock: Eth1Block) =
   eth1Chain.blocks.addLast newBlock
   eth1Chain.blocksByHash[newBlock.voteData.block_hash.asBlockHash] = newBlock
 
+template hash*(x: Eth1Data): Hash =
+  hash(x.block_hash.data)
```
Comment (protolambda) [inline beacon_chain/eth1_monitor.nim:204]: what is happening here? Hash of the Eth1Data for uniqueness? If this is untrusted eth1data from elswhere, e.g. the ongoing votes in the beacon state, then Eth1Data objects can have the same block hash but different deposit data roots and deposit counts. Don't count those as the same thing.

Comment (arnetheduck) [reply beacon_chain/eth1_monitor.nim:204]: this is an insecure hash used for hash table / dictionary lookups (`Table`), though good point as it'll cause collisions

**Comment by zah** [comment]
> lots of lengthy lockups when processing the eth1 blocks - also, while it's processing pre-genesis blocks, it returns a pre-genesis eth1 vote

We'll start baking a merkleizer corresponding to the pre-genesis deposits, so they will never be processed in the future.

**Pull request merged**